feat: add battery monitoring and detailed power state getter

Ignigena opened the pull request:
## Description

Added `getPowerState()` that allows for fetching more detailed information about the power on the device. In addition to returning the battery level it includes information about the battery state (full, charging, etc.) and whether the system is operating in low power mode.

Additionally, this adds a few events that can be subscribed to in order to be notified when the battery level changes, state changes (for example, if the device is unplugged), or if the battery drops below 20%.

These additional capabilities are very helpful in our case as we want to trigger certain things within our app when an app intended for kiosk use is unplugged and starts depleting battery. It's also useful to know if the battery is low or the user has switched on low power mode so that we can defer certain actions until the battery is above a certain threshold.

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| iOS     |    ✅     |
| Android |    ❌     |
| Windows |    ❌     |

## Checklist

<!-- Check completed item: [X] -->

* [x] I have tested this on a device/simulator for each compatible OS
* [x] I added the documentation in `README.md`.
* [ ] I mentionned this change in `CHANGELOG.md`.

mikehardy commented:
@Ignigena are you still using your changes in production? I could try to help these merge...there are similar capabilities on Android even so it could be a nice symmetrical implementation

Ignigena commented:
We are using this in production although haven’t upgraded the library since opening the PR. Would love to get this merged in if the feature is still missing and there’s desire to do so :)

mikehardy commented:
Hi there! The feature is still missing - the closest thing is simply having functional iOS status #404 (but that was just a doc update, really) and on android side an implementation of isBatteryCharging #514 

Looks like you'll need to rebase against master and fix some conflicts, but then after that maybe we can get it in shape for merge

Ignigena commented:
Sorry for the delay. Everything's been rebased and should be ready for next steps!

Ignigena commented on README.md:
```diff
@@ -652,6 +652,24 @@ const phoneNumber = DeviceInfo.getPhoneNumber();
 
 ---
 
+### getPowerState()
```
Just added it. I didn't put anything under the "available since" column since I'm not sure what release this will actually land in.

On the Android note it does look like [they have a similar API](https://developer.android.com/training/monitoring-device-state/battery-monitoring). Unfortunately I don't have a physical Android device to test on, so while I could take a stab at it I'm not sure I'd be the best one to push it across the finish line :)

mikehardy replied:
this looks nice, and we can't all do everything, normally, happy to have an iOS dev, and I need the Java stuff I'll just have to code it up...

Ignigena commented on README.md:
````diff
@@ -964,6 +982,50 @@ Returns a list of supported processor architecture version
+
+```js
+deviceInfoEmitter.addListener('batteryLevelDidChange', level => {
+  // 0.759999
````
Really good point. It's not exactly straightforward and a bit extra clarity certainly can't hurt. I added two lines that show how to set up the emitter for `RNDeviceInfo`. The usual caveats apply about using a single emitter, cleaning up if this is done inside a mounted component, etc etc.

mikehardy replied:
This is a big help to me (and hopefully useful to future devs), thanks

Ignigena commented on deviceinfo.js.flow:
```diff
@@ -43,6 +43,7 @@ declare module.exports: {
   getTotalDiskCapacity: () => number,
   getFreeDiskStorage: () => number,
   getBatteryLevel: () => Promise<number>,
+  getPowerState: () => Promise<object>,
```
Had no idea this file existed :) Just added it.

On that note, is the web polyfill actually being used and is there interest in having it actually return anything except blank data? While `navigator.getBattery` on the web side has pretty lousy browser support it shouldn't be too hard to actually get some sane values for some of the other methods either with direct browser functions or basic sniffing like we're doing in some places on the iOS side anyway.

mikehardy replied:
I think I mentioned this, but the web polyfill is specifically mentioned as a courtesy. If I start seeing PRs to implement in it, I'll take interest, in the meantime it's just to allow them to compile

Ignigena commented on ios/RNDeviceInfo/RNDeviceInfo.m:
```diff
@@ -376,14 +402,48 @@ - (NSDictionary *)constantsToExport
+             @"batteryState": @"full",
+#else
+             @"batteryLevel": @([UIDevice currentDevice].batteryLevel),
+             @"batteryState": [@[@"unknown", @"unplugged", @"charging", @"full"] objectAtIndex: [UIDevice currentDevice].batteryState],
```
Just added a warning log if running in dev mode and on an actual device (IIRC from testing the simulator always returns `unknown` so this could be confusing if you've already enabled the battery monitoring mode)

mikehardy replied:
I see the warning, that is helpful. I do much of my testing on an iOS emulator so I'll notice, if you can confirm the emulator always returns unknown since it is a dev notice anyway maybe you could add that?  "Emulators typically return unknown every time." or whatever. The pain now will avoid 20 duplicate issues over 2 years I bet

mikehardy replied:
The emulator does return unknown in my testing, if you accept my PR to your fork (which will then push my suggested warning changes into this PR) then this will be resolved

mikehardy replied:
thanks, resolved this one

Ignigena commented:
Also just realized that `isBatteryCharging` was only implemented on Android. Since the same information comes out of the `powerState` dictionary I added a new method which brings compatibility for iOS.

machour commented on ios/RNDeviceInfo/RNDeviceInfo.m:
```diff
@@ -421,8 +421,8 @@ - (NSDictionary *)powerState
     if ([UIDevice currentDevice].batteryState == UIDeviceBatteryStateUnknown) {
-        RCTLogWarn(@"Battery state is `unknown` which could mean battery monitoring is disabled." +
-                    "You may need to enable monitoring with `[UIDevice currentDevice].batteryMonitoringEnabled = TRUE`");
+        RCTLogWarn(@"Battery state is `unknown` which could mean battery monitoring is disabled. "
```
isn't the `+` needed for concatenation?

Ignigena replied:
That's what I thought too...

<img width="872" alt="Screen Shot 2019-03-26 at 11 56 59 AM" src="https://user-images.githubusercontent.com/4356436/55017234-4d410c80-4fbe-11e9-870b-be611d158953.png">

Ignigena replied:
Also see: https://github.com/facebook/react-native/blob/master/React/Base/RCTModuleData.mm#L65

mikehardy commented:
Nice! I see @machour  has already taken look. I will review this when I get a moment later tonight or tomorrow morning (it's afternoon where I am now). I especially appreciate adding the android bit :-). I'd ignore the web part for now as the polyfill is just to get it to compile and I haven't seen much interest (in the form of PRs, indicating actual interest) in it providing real values.

mikehardy commented:
When I pulled it and tested it locally with additions to the example (see the PR I opened on your fork) I saw these things

one is that the object keys seem odd, and the other is that I think we should provide a default empty implementation of the listener for the emitter to avoid the warnings? Then maybe an example of the emitter listener in the App.js of the example?

But it all worked! I definitely had data :-)

![IMG_20190326_234719__01__01](https://user-images.githubusercontent.com/782704/55051804-a93e7c00-5024-11e9-99fb-0d0cb56ef836.jpg)
![IMG_20190326_234737__01](https://user-images.githubusercontent.com/782704/55051805-a93e7c00-5024-11e9-824b-8324661948d3.jpg)

machour commented:
The keys from your screenshot look like a promise internal thing. How did you call the code ?

mikehardy commented:
> The keys from your screenshot look like a promise internal thing. How did you call the code ?

@machour nailed it! Called it like so (I did a PR on this PR over on his repo to add the example snippet) https://github.com/photomadic/react-native-device-info/pull/1/files#diff-94034572e4cba4c48295b32b8ca49e68

Forgot the await before the call. 🤦‍♂️ 

So, I think this is just my PR into the fork (to add the example, but *with* the await) then default emitters then I've actually seen it work and it's a great feature ?

mikehardy commented on ios/RNDeviceInfo/RNDeviceInfo.m:
```diff
@@ -376,14 +402,60 @@ - (NSDictionary *)constantsToExport
+- (void)powerStateDidChange:(NSNotification *)notification
+{
+    [self sendEventWithName:@"powerStateDidChange" body:self.powerState];
+}
```
When running this I saw a warning about emitting events without a listener. I originally thought adding a default listener was an idea to fix it but I read more, and the thing to do is to actually toggle whether you listen to system events and emit them to javascript based on whether you have listeners or not - the idea being that you don't do work (and spend device energy) if you don't have to. So you should subscribe/unsubscribe to the iOS service based on whether you have listeners. Explanation with code example here: https://facebook.github.io/react-native/docs/native-modules-ios.html#optimizing-for-zero-listeners

mikehardy replied:
perfect

mikehardy commented:
Starting point for Android implementation - idea would be to register for the various power-related intents in a BroadcastReceiver:
ACTION_BATTERY_LOW
ACTION_BATTERY_OKAY
- sent once after ACTION_BATTERY_LOW (allegedly)
ACTION_BATTERY_CHANGED
- must explicitly register https://developer.android.com/reference/android/content/Intent#ACTION_BATTERY_CHANGED
ACTION_POWER_CONNECTED
ACTION_POWER_DISCONNECTED

Careful to handle device-without-battery cases (they changed the default return values for those devices between API27 and API28 https://source.android.com/devices/tech/power/batteryless )

Then in the broadcast receiver emit them similar to the iOS implementation, using the info here:
https://developer.android.com/reference/android/os/BatteryManager.html

Android publishes a great deal more information though, which might be nice to expose as a sub-object in the JSON - battery cycles and current draw, time to charge / time to drain etc

mikehardy commented:
Final stretch on this PR - if you can toggle whether to emit events or not based on whether you have subscribers are not we have the highest quality PR possible :-)

Ignigena commented:
Alrighty, just added the extra boolean to track whether event listeners have been registered and emit accordingly. Rebased against the latest changes in `master`. Let me know if you see anything else that should be addressed as part of this PR!

mikehardy commented on refresh-example.sh:
```diff
@@ -37,7 +35,7 @@ sed -i -e 's/INTERNET" \/>/INTERNET" \/><uses-permission android:name="android.p
 
 # Patch the AppDelegate for iOS battery level
-sed -i -e 's/return YES;/\[UIDevice currentDevice\].batteryMonitoringEnabled = true; return YES;/' ios/example/AppDelegate.m
+sed -i -e $'s/return YES;/\[UIDevice currentDevice\].batteryMonitoringEnabled = true;\\\n  return YES;/' ios/example/AppDelegate.m
```
Yeah - you saw that too :-) - I saw it and fixed it while testing this

mikehardy commented:
I made one minor adjustment to the code and then doc updates as a PR to your branch a second time. It looks *perfect* to me with the tiny code change. Let me know what you think and merge if it's good, then I'll hit the squash+merge button.

This is a really nice feature - thanks!

machour commented on README.md:
```diff
@@ -656,6 +656,7 @@ const phoneNumber = DeviceInfo.getPhoneNumber();
 ### getPowerState()
 
 Gets the power state of the device including the battery level, whether it is plugged in, and if the system is currently operating in low power mode.
+Displays a warning on iOS if battery monitoring not enabled, or if attempted on an emulator (where montioring is not possible)
```
s/montioring/monitoring/

mikehardy replied:
Nice catch!

mikehardy commented:
@Ignigena you're a champion for working through this with me, I feel like it's a really high-quality piece of code, thanks! Also, my family lives in Austin :-), everyone loves to share their brisket recommendations right? So if you get a chance and you haven't tried it give the [Micklethwaite](https://craftmeatsaustin.com/) trailer a try ;-). Cheers

mikehardy commented:
react-native-device-info 1.4.0 published 🎉
